File: NeetCode/Stack/Asteroid_Collision/asteroid__collision.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []

        for asteroid in asteroids:
            alive = True

            while alive and stack and stack[-1] > 0 and asteroid < 0:
                if stack[-1] < -asteroid:
                    # Existing asteroid is smaller.
                    stack.pop()
                elif stack[-1] == -asteroid:
                    # Both are destroyed.
                    stack.pop()
                    alive = False
                else:
                    # Incoming asteroid is smaller.
                    alive = False

            if alive:
                stack.append(asteroid)

        return stack


if __name__ == "__main__":
    sol = Solution()
    asteroids = [-2, -1, 1, 2]
    print(sol.asteroidCollision([2, 4, -4, -1]))
    print(sol.asteroidCollision([7, -3, 5]))
    print(sol.asteroidCollision([5, 5]))
    print(sol.asteroidCollision([-2, -1, 1, 2]))
    print("Running Solution...")

File: NeetCode/Stack/Asteroid_Collision/Asteroid_Collision.cpp
// Asteroid_Collision.cpp
//
// Compile locally with:
// clang++ Asteroid_Collision.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    vector<int> asteroidCollision(vector<int> &asteroids)
    {
        vector<int> stk;
        for (auto &asteroid : asteroids)
        {
            bool alive = true;
            while (alive && !stk.empty() && stk.back() > 0 && asteroid < 0)
            {
                if (stk.back() < -asteroid)
                {
                    stk.pop_back();
                }
                else if (stk.back() == -asteroid)
                {
                    stk.pop_back();
                    alive = false;
                }
                else
                {
                    alive = false;
                }
            }
            if (alive)
            {
                stk.push_back(asteroid);
            }
        }
        return stk;
    }
};

int main()
{
    cout << "Running Asteroid_Collision.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}
